extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/DataServerQueryHandler.java

```diff
@@ -195,7 +200,35 @@ public <RowType, QueryType> DataServerQueryResult<RowType> fetchRowsFromDataServ
       }
     }
 
-    return new DataServerQueryResult<>(yielders, handedOffSegments, dataSource);
+    return new DataServerQueryResult<>(yielders, handedOffSegments, dataSourceName);
+  }
+
+  /**
+   * Transform the {@link InputNumberDataSource}, which are only understood by MSQ tasks, back into a
+   * {@link TableDataSource}.
+   */
+  private DataSource transformDatasource(DataSource dataSource)
+  {
+    if (dataSource instanceof InputNumberDataSource) {
+      InputNumberDataSource numberDataSource = (InputNumberDataSource) dataSource;
+      if (numberDataSource.getInputNumber() == inputNumber) {
+        return new TableDataSource(dataSourceName);
+      } else {
+        throw DruidException.forPersona(DruidException.Persona.USER)
+                            .ofCategory(DruidException.Category.INVALID_INPUT)
+                            .build("Unknown InputNumberDataSource datasource with number[%s]. Queries with realtime sources "
```

Comment:
This is an end-user-facing error message, it shouldn't start with `Unknown InputNumberDataSource datasource with number[1].` because that doesn't mean much to an end-user. The term `stage outputs` isn't very clear either. Please rewrite it to be more useful to the end user. It should describe concepts they would be familiar with from the SQL side, such as mentioning that broadcast joins are not currently supported with realtime data.

Also, could this situation (an `InputNumberDataSource` for some non-base datasource) happen in any case other than joins? If so the error message should reflect that. Maybe unions?